libdnf/sack/query.cpp

```diff
@@ -1745,31 +1768,47 @@ Query::Impl::filterLatest(const Filter & f, Map *m)
         if (keyname == HY_PKG_LATEST_PER_ARCH) {
             solv_sort(samename.elements, samename.count, sizeof(Id),
                       filter_latest_sortcmp_byarch, pool);
+        } else if (keyname == HY_PKG_LATEST_PER_ARCH_BY_PRIORITY) {
+            solv_sort(samename.elements, samename.count, sizeof(Id),
+                      filter_latest_sortcmp_byarch_bypriority, pool);
         } else {
             solv_sort(samename.elements, samename.count, sizeof(Id),
                       filter_latest_sortcmp, pool);
         }
 
         Solvable *considered, *highest = 0;
+        bool make_block = 1;
         int start_block = -1;
         int i;
         for (i = 0; i < samename.count; ++i) {
             Id p = samename.elements[i];
             considered = pool->solvables + p;
-            if (!highest || highest->name != considered->name ||
-                ((keyname == HY_PKG_LATEST_PER_ARCH) && highest->arch != considered->arch)) {
+            if (!highest ||
+                highest->name != considered->name ||
+                ((keyname == HY_PKG_LATEST_PER_ARCH || keyname == HY_PKG_LATEST_PER_ARCH_BY_PRIORITY) &&
+                highest->arch != considered->arch)) {
                 /* start of a new block */
                 if (start_block == -1) {
                     highest = considered;
                     start_block = i;
                     continue;
                 }
-                add_latest_to_map(pool, m, &samename, start_block, i, latest);
+                if (make_block) {
+                    add_latest_to_map(pool, m, &samename, start_block, i, latest);
+                }
+                else {
+                    make_block = 1;
+                }
                 highest = considered;
                 start_block = i;
+            } else if (keyname == HY_PKG_LATEST_PER_ARCH_BY_PRIORITY &&
+                highest->repo->priority != considered->repo->priority &&
+                make_block) {
+                add_latest_to_map(pool, m, &samename, start_block, i, latest);
+                make_block = 0;
             }
         }
-        if (start_block != -1) {
+        if (start_block != -1 && make_block) {
```

Comment:
Please can you add a comment "// Add last block to the map"? Or something else?

Reply:
I added the comment and two more describing what is going on. I hope they are correct.